pkg/cachemanager/cachemanager_test.go

```diff
@@ -0,0 +1,373 @@
+	"github.com/open-policy-agent/gatekeeper/v3/test/testutils"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+)
+
+var cfg *rest.Config
+
+func TestMain(m *testing.M) {
+	testutils.StartControlPlane(m, &cfg, 2)
+}
+
+func unitCacheManagerForTest(t *testing.T) (*CacheManager, context.Context) {
+	cm, _, ctx := makeCacheManager(t)
+	return cm, ctx
+}
+
+func makeCacheManager(t *testing.T) (*CacheManager, client.Client, context.Context) {
+	mgr, wm := testutils.SetupManager(t, cfg)
+	c := testclient.NewRetryClient(mgr.GetClient())
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	cfClient := &FakeCfClient{}
+	tracker, err := readiness.SetupTracker(mgr, false, false, false)
+	require.NoError(t, err)
+	processExcluder := process.Get()
+	processExcluder.Add([]configv1alpha1.MatchEntry{
+		{
+			ExcludedNamespaces: []wildcard.Wildcard{"kube-system"},
+			Processes:          []string{"sync"},
+		},
+	})
+	events := make(chan event.GenericEvent, 1024)
+	w, err := wm.NewRegistrar(
+		"test-cache-manager",
+		events)
+	require.NoError(t, err)
+
+	cacheManager, err := NewCacheManager(&Config{
+		CfClient:         cfClient,
+		SyncMetricsCache: syncutil.NewMetricsCache(),
+		Tracker:          tracker,
+		ProcessExcluder:  processExcluder,
+		WatchedSet:       watch.NewSet(),
+		Registrar:        w,
+		Reader:           c,
+		GVKAggregator:    aggregator.NewGVKAggregator(),
+	})
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		ctx.Done()
+	})
+
+	t.Cleanup(func() {
+		cancelFunc()
+	})
+
+	testutils.StartManager(ctx, t, mgr)
+
+	return cacheManager, c, ctx
+}
+
+// TestCacheManager_wipeCacheIfNeeded.
+func TestCacheManager_wipeCacheIfNeeded(t *testing.T) {
+	cacheManager, ctx := unitCacheManagerForTest(t)
+	cfClient, ok := cacheManager.cfClient.(*FakeCfClient)
+	require.True(t, ok)
+
+	// seed one gvk
+	configMapGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "ConfigMap"}
+	cm := unstructuredFor(configMapGVK, "config-test-1")
+	_, err := cfClient.AddData(ctx, cm)
+	require.NoError(t, err, "adding ConfigMap config-test-1 in cfClient")
+
+	podGVK := schema.GroupVersionKind{Group: "", Version: "v1", Kind: "Pod"}
+	require.NoError(t, cacheManager.gvksToSync.Upsert(aggregator.Key{Source: "foo", ID: "bar"}, []schema.GroupVersionKind{podGVK}))
+
+	cacheManager.gvksToDeleteFromCache.Add(configMapGVK)
+	cacheManager.wipeCacheIfNeeded(ctx)
+
+	require.False(t, cfClient.HasGVK(configMapGVK))
+	require.ElementsMatch(t, cacheManager.gvksToSync.GVKs(), []schema.GroupVersionKind{podGVK})
+}
+
+// TestCacheManager_syncGVKInstances tests that GVK instances can be listed and added to the cfClient client.
+func TestCacheManager_syncGVKInstances(t *testing.T) {
```

Comment:
This is a lot of test for not much coverage.  We really just get a happy path of syncGVK, which we'd get from an e2e test.

Personally I would delete this test.

If you'd like to keep it, I'd make it a table driven test that examines the various pathways of the conditional logic (since that's what we really want to confirm works correctly when we test):

1.  Objects of a watched GVK end up in the cache
2. Objects of a non-watched GVK do not end up in the cache

The other permutations (like process exclusion) are tested at the level of AddObject, so they can be omitted from this test.